File: 0653-two-sum-iv-input-is-a-bst/0653-two-sum-iv-input-is-a-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean findTarget(TreeNode root, int k) {
        return findTarget(root, k, new HashSet<Integer>());
    }
    
    public boolean findTarget(TreeNode root, int k, Set<Integer> set){
        if(root == null){
            return false;
        }
        
        if(set.contains(k - root.val)){
            return true;
        } 
        
        set.add(root.val);
        
                
        return findTarget(root.left, k, set) || findTarget(root.right, k, set);
    }
}

File: Easy/0653-two-sum-iv-input-is-a-bst/0653-two-sum-iv-input-is-a-bst.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} k
 * @return {boolean}
 */
var findTarget = function(root, k) {
    const set = new Set();
    
    function recursive(root, k, set){
        if(root === null){
            return false;
        }
        
        if(set.has(k - root.val)){
            return true;
        }
        
        set.add(root.val);
        
        return recursive(root.left, k, set) || recursive(root.right, k, set);
        
    }
    
    return recursive(root, k, set);
};